Serverless Page within APM Docs (#8132)
DataDog/documentation: content/en/tracing/serverless_functions/_index.md

---
title: Trace Serverless Functions
kind: documentation
description: 'Send traces from your Serverless Functions to Datadog'
further_reading:
    - link: 'serverless'
      text: 'Serverless Monitoring with Datadog'
---
{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

In hybrid architectures, Serverless functions can form an important core of requests your application processes. Including these functions in your Distributed Traces can be critical for detecting performance bottlenecks and service outages in your distributed systems.

> ## Install the AWS Integration

Reviewer: We should change this section to be like "Choose your tracing library" since that's what this is really about

Reviewer: or something more generall like "Serverless Disributed Tracing"

> As a first step, make sure to [install the AWS integration][1] before proceeding.

Reviewer: We can move this to when we're explicitly describing the steps to install AWS X-Ray

> Next, depending on your language and configuration, choose between using the Datadog Lambda Library or Amazon X-Ray for your traces.

Reviewer: We can add a sentence after here saying something like: if you are looking to trace across functions and hosts, you need to set up both.

Reviewer: and we can link to this picture to explain what we mean

https://datadog-docs.imgix.net/images/integrations/amazon_lambda/lambda_tracing.cf509d6b0ca3b5dabb86dbaf7a13f906.png?fit=max&auto=format

{{< img src="tracing/serverless_functions/LambdaLibraryXRay.png" alt="Lambda Library or X-Ray"  style="width:100%;">}}

> ## Use the Datadog Lambda Library

Reviewer: ```suggestion
## Use the Datadog APM
```

Reviewer: I also think we should be more explicit in this section in being like "these docs live elsewhere"

Reviewer: In general, on the UX side this section doesn't feel well highlighted, and really easy to skip over.

> For [Java][2], [Node][3], [Python][4], [Ruby][5] and [Go][6], the [Datadog Lambda Library][7] is recommended and has several options available for installation.

Reviewer: We don't support Datadog tracing for .NET, Go and Java right now. This might be unclear here, because some customers probably think Lambda Library == tracing.

> ## Use the X-Ray Integration

Reviewer: ```suggestion
## Enable AWS X-Ray
```

```python
from aws_xray_sdk.core import xray_recorder
from aws_xray_sdk.core import patch_all
patch_all()
```

Note that tracing `aiohttp` requires [specific instrumentation][2].

For further configuration, creating subsegments, and recording annotations, consult the [X-Ray Python docs][3].

[1]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-python-patching.html
[2]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-python-httpclients.html
[3]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-python.html
{{% /tab %}}
{{% tab "Go, Ruby, Java, .NET" %}}

For all other runtimes, consult the X-Ray SDK docs:

- [X-Ray SDK for Go][1]
- [X-Ray SDK for Ruby][2]
- [X-Ray SDK for Java][3]
- [X-Ray SDK for .NET & Core][4]

[1]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-go.html
[2]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-ruby.html
[3]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-java.html
[4]: https://docs.aws.amazon.com/en_pv/xray/latest/devguide/xray-sdk-dotnet.html
{{% /tab %}}
{{< /tabs >}}

> ### Data Collected

Reviewer: We can keep this on the AWS X-Ray integration page

The AWS X-Ray integration pulls in trace data from AWS, and does not collect any metrics or logs.

## Tracing across AWS Lambda and hosts

> When applicable, Datadog merges AWS X-Ray traces with native Datadog APM traces. This means that your traces will show the complete picture of requests that cross infrastructure boundaries, whether it be AWS Lambda, containers, on-prem hosts, or managed services.

Reviewer: I think we should provide another reminder that in many cases, using either Datadog APM or AWS X-Ray is sufficient.

I would also add a note saying that enabling both could lead to higher costs.

Author: is this now resolved by the above changes?

1. Enable the [AWS X-Ray integration][13] for tracing your Lambda functions.

> 2. Add the [Datadog Lambda Library](#Datadog-Lambda-Library) to your Lambda functions.

Reviewer: We can be clearer here and say "Set up Datadog APM on your Lambda functions" or something like that.

Reviewer: I think the goal being that Installing the Lambda Library is implementation for this larger goal.

> 3. [Set up Datadog APM][14] on your hosts and container-based infrastructure.

Reviewer: I think we should provide some code samples for how this looks at the end of the day, specifying which parts of the code belong to Datadog APM, and which are part of X-Ray.

**Note**: For X-Ray and Datadog APM traces to appear in the same flame graph, all services must have the [same `env` tag](#the-env-tag).

{{< img src="integrations/amazon_lambda/lambda_host_trace.png" alt="trace of a request from a host to a Lambda function" >}}

> ### Organizing your infrastructure with tags

Reviewer: I'm not sure these instructions are the best practice for tagging your functions with X-Ray? What do you think?

Reviewer: FWIW we're looking to provide more information on tagging on the serverless side.

---
title: Trace Serverless Functions
kind: documentation
description: 'Send traces from your Serverless Functions to Datadog'
further_reading:
    - link: 'serverless'
      text: 'Serverless Monitoring with Datadog'
---
{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

In hybrid architectures, serverless functions are a key part of your application. Including these functions in your Distributed Traces can be critical for detecting performance bottlenecks and service outages in your distributed systems.

## Choose your Tracing Library

Depending on your language and configuration, choose below between using the Datadog Lambda Library or AWS X-Ray for your traces and follow the respective instructions to complete your setup.

{{< partial name="apm/apm-serverless.html" >}}

## Augment AWS X-Ray Tracing with Datadog APM

{{< img src="integrations/amazon_lambda/lambda_tracing.png" alt="architecture diagram for tracing AWS Lambda with Datadog" >}}

> There are some scenarios in which AWS X-Ray Tracing will be used side-by-side with Datadog APM.  Note that this will have an impact on your bill, and if you are unsure whether to use Datadog APM or AWS X-Ray, please reach out to [our support team][1] to discuss, or read about the [differences](#choose-your-Tracing-Library).

Author: Tried something - let me know what you think

---
title: Trace Serverless Functions
kind: documentation
description: 'Send traces from your serverless functions to Datadog'
further_reading:
    - link: 'serverless'
      text: 'Serverless Monitoring with Datadog'
---
{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

In hybrid architectures, serverless functions are a key part of your application. Including these functions in your distributed traces can be critical for detecting performance bottlenecks and service outages in your distributed systems.

## Choose your Tracing Library

> Depending on your language and configuration, choose between setting up the Datadog Lambda Library or AWS X-Ray for your traces:

Reviewer: ```suggestion
Depending on your language and configuration, choose between setting up the Datadog APM or AWS X-Ray for your traces:
```

> {{< partial name="apm/apm-serverless.html" >}}

Reviewer: ```suggestion
```

---
title: Trace Serverless Functions
kind: documentation
description: 'Send traces from your serverless functions to Datadog'
further_reading:
    - link: 'serverless'
      text: 'Serverless Monitoring with Datadog'
---
{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

In hybrid architectures, serverless functions are a key part of your application. Including these functions in your distributed traces can be critical for detecting performance bottlenecks and service outages in your distributed systems.

## Choose your Tracing Library

> Depending on your language and configuration, choose between setting up Datadog APM or AWS X-Ray for your traces:

Reviewer 2: ```suggestion
Depending on your language and configuration, choose between setting up Datadog APM or AWS X-Ray for your traces. Click the link in the appropriate column heading to see set-up instructions for the solution that suits you:
```

---
title: Trace Serverless Functions
kind: documentation
description: 'Send traces from your Serverless Functions to Datadog'
further_reading:
    - link: 'serverless'
      text: 'Serverless Monitoring with Datadog'
---
{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

In hybrid architectures, Serverless functions can form an important core of requests your application processes. Including these functions in your Distributed Traces can be critical for detecting performance bottlenecks and service outages in your distributed systems.

## Choose your Tracing Library

Depending on your language and configuration, choose between using the Datadog Lambda Library or AWS X-Ray for your traces.

{{< img src="tracing/serverless_functions/LambdaLibraryXRay.png" alt="Lambda Library or AWS X-Ray"  style="width:100%;">}}

**Note:** If you are tracing across lambda functions and hosts, you will need to use **both** Datadog APM and AWS X-Ray.

> ## Enable Datadog APM

Reviewer 3: ```suggestion
## Recommended: Enable Datadog APM
```

Reviewer 3: @andrewardito can we do something like this to guide customers along the happy path? If I'm an engineer with two hours to set up APM on my service, I'm not sure I'd be able to figure this out.

> Instructions for enabling Datadog APM on your lambda functions are contained within the [Serverless][1] section of our documentation.  You will also need to [install the AWS integration][2].

Reviewer 3: ```suggestion
Instructions for enabling Datadog APM on your AWS Lambda functions are contained within the [Serverless][1] section of our documentation. You will also need to [install the AWS integration][2].
```

## Enable AWS X-Ray

> Instructions for enabling AWS X-Ray on your lambda functions are contained within the [AWS X-Ray Tracing][3] section of our documentation.  You will also need to [install the AWS integration][2].

Reviewer 3: ```suggestion
Instructions for enabling AWS X-Ray on your lambda functions are contained within the [AWS X-Ray Tracing][3] section of our documentation. You will also need to [install the AWS integration][2].
```

## Augment AWS X-Ray Tracing with Datadog APM

{{< img src="integrations/amazon_lambda/lambda_tracing.png" alt="architecture diagram for tracing AWS Lambda with Datadog" >}}

> There are some scenarios in which AWS X-Ray Tracing will be used side-by-side with Datadog APM.  Note that this will have an impact on your bill, and if you are unsure whether to use Datadog APM or AWS X-Ray, please reach out to [our support team][4] to discuss, or read about the [differences](#choose-your-Tracing-Library)

Reviewer 3: ```suggestion
There are some scenarios in which AWS X-Ray Tracing will be used side-by-side with Datadog APM.  Note that this will have an impact on your bill, and if you are unsure whether to use Datadog APM or AWS X-Ray, please reach out to [our support team][4] to discuss, or read about the [differences](#choose-your-Tracing-Library).
```

> ### Tracing in a serverless-first environment

Reviewer 3: I like the intention here, but as a user skimming the doc, this catches my eye. The whole X-Ray section being so long and detailed on this page makes it easy to miss the happy path that we highlight under the **Datadog APM** header

Author: Besides adding Recommended to that section, anything else we should do here?

> AWS X-ray is both a backend AWS service and a set of client libraries. The service gives you an Invocation span for your lambda functions and traces across api gateways and message queues.

Reviewer 3: ```suggestion
AWS X-Ray is both a backend AWS service and a set of client libraries. The service gives you an Invocation span for your AWS Lambda functions and traces across Amazon API Gateways and message queues.
```

The client libraries trace your integrations in your code. If both of these are important to trace and visualize in your flame graphs, follow the below two steps.

1. Enable the [AWS X-Ray integration][3] for tracing your Lambda functions.
2. [Set up Datadog APM][5] on your Lambda functions.

### Tracing across AWS Lambda and hosts

When applicable, Datadog merges AWS X-Ray traces with native Datadog APM traces. This means that your traces will show the complete picture of requests that cross infrastructure boundaries, whether it be AWS Lambda, containers, on-prem hosts, or managed services.

1. Enable the [AWS X-Ray integration][3] for tracing your Lambda functions.
2. [Set up Datadog APM][6] on your hosts and container-based infrastructure.

**Note**: For X-Ray and Datadog APM traces to appear in the same flame graph, all services must have the [same `env` tag](#the-env-tag).

{{< img src="integrations/amazon_lambda/lambda_host_trace.png" alt="trace of a request from a host to a Lambda function" >}}

### Organizing your Serverless infrastructure with tags

> Any [tag][7] applied to your Lambda function automatically becomes a new dimension on which your can slice and dice your traces.

Reviewer 3: ```suggestion
Any [tag][7] applied to your AWS Lambda function automatically becomes a new dimension on which your can slice and dice your traces.
```

Tags are especially powerful for consistency across the Datadog platform, which has [first-class support][8] for the `env` and `service` tags.

**Note**: If you are tracing with Datadog APM, set the parameter `DdFetchLambdaTags` to `true` on the forwarder CloudFormation stack to ensure your traces are tagged with the resource tags on the originating Lambda function. Lambda function resource tags are automatically surfaced to X-Ray traces in Datadog without any additional configuration.

#### The env tag

Use `env` to separate out your staging, development, and production environments. This works for any kind of infrastructure, not just for your serverless functions. As an example, you could tag your production EU Lambda functions with `env:prod-eu`.

> By default, Lambda functions are tagged with `env:none` in Datadog. Add your own tag to override this.

Reviewer 3: ```suggestion
By default, AWS Lambda functions are tagged with `env:none` in Datadog. Add your own tag to override this.
```
